### ISSUE TITLE
[Android] Topic file gets resetted over time

### ISSUE THREAD

**Issue by Tobi-La** [bwilcox-1234/ChatScript#114]
Hi, 
i am using Chatscript as part of my android game.
On one of my test devices I have the problem that my user topic file gets recreated and so everything the user has achieved so far gets deleted.

Currently I have no idea how to further inspect the issue, I have no idea why chatscript is doing this.

The issue is always happening when my app is operating in the background. It is calling Chatscript twice a second to check if there is any update in the data chatscript provides.

Here is what I inspected what happens when the issue occurs:

- The USERS/topic_user_bot.txt is recreated (the creation date in the second line is refreshed and everything saved in the topic file gets deleted ( user facts, variables, history))
- There is no entry in the LOGS/bugs.txt
- the USERS/log_user.txt is not reset, but I couldn't recognize anything special in there, since it only logs IO

Is there another log I can activate to check what the engine is doing?

Thanks in advance,
Tobi

**Comment by bwilcox-1234** [comment]
So you are saying this does NOT happen on other of your test devices. And presumably your code implementation is the same on both. So it's not universal engine behavior (nor should it be).  A topic file will get recreated if it has a faulty read.  Are you saying this never happens when your app is in the foreground?

**Comment by Tobi-La** [comment]
I only have two devices, it happens on only one of them, yes.
It's the excact same version of the app installed on both devices.
It didn't occur yet when the app is in foreground, but that could be due to the fact that I didn't test to have the app in foreground doing nothing for a huge amount of time yet. I will test that and report if this would be helpful. On the other hand the way chatscript get called is the same in background or foreground mode of the app.

Is Chatscript doing some logging for the case that it can't read the topic file and recreates it? Maybe it has an error code too why the file isn't readable, I think that would be very helpful!

**Comment by bwilcox-1234** [comment]
if it only happens on one device, then the issue is not chatscript engine per se.  But you will want to test whether this problem can happen in foreground also.   If :trace user  is on, it will log that it generated a new user. Regardless, if user topic file read fails in some part, it reports into LOGS/BUGS.txt what failed

**Comment by Tobi-La** [comment]
I let the display of the device turned on all night to let the app stay in foreground. The issue didn't occur. Pushed the app to the background, the issue occured 20 minutes later. So it really seems to be something the device is doing when the app is in background.

Unfortunately there is really nothing reported in the LOGS/BUGS.txt. And I am not sure yet how to turn on :trace user inside my android app, cause it will break the messages the app expects. Would it log some helpful / new information with it, or would it just log that it recreated the userfile?

Is there another way to get the log /  error code why the file couldn't be opened?

**Comment by bwilcox-1234** [comment]
chatscript directly recognizes : commands, like :trace all, so regardless of what your app is expecting, will it not accept :trace all to turn it on?

**Comment by Tobi-La** [comment]
Yes it does, but the way chatscript is communicating with my app is entirely by json oob. I could enhance my app to handle data that is not json somehow, like write it to a logfile, but that wouldn't be as easy as it might sound. So I wanted to make sure that the information we get would help solving the problem or not.
If there are any information using :trace that could help me I will of course do it, but from what I tested with the standalone version it logs only data of my own chatscript scriptcode. I already know what happens there because I programmed some kind of own logging behavior:

To one point everything works normal, then with the next call of chatscript it just starts all over again:

Log (very much simplified because it logs very app specific infos:

Chatscript call 1: bot outputmacro gets called + normal expected behavior
....
Chatscript call x (lets say 2347) : normal expected behavior
Chatscript call x+1 (2348) bot outputmacro gets called again, everything starts like call 1 again

**Comment by bwilcox-1234** [comment]
offhand I dont see any existing data that will inform us why file/io fails coming out of background mode on a specific device. I presume that a simple file open or file read call fails, returning a code which CS doesnt keep around at present. I might guess that the problem is not in read but in file open itself. Once opened, I presume reading wouldnt fail.  What you'd want is to modify CS source to keep around an errno value and pass that back in oob.

**Comment by Tobi-La** [comment]
Alright, I analyzed the problem:

Chatscript is trying to open the user topicfile in the FopenReadWritten in os.cpp. the fopen fails with errno 4: Interrupted system call

As you said, Chatscript is not interested in this error code atm and handles this case as if the file wasn't there.

Now the final question for this topic:
In my opinion it would make sence to check the error code fopen is producing if it fails, and only continue if it fails with errno 2: no such file or directory.
In every other case I think it is wrong to simply continue, because the file may be there and we are risking data loss, not only in my case. In my opinion a retry of fopen and / or canceling the operation with an error code would be better than overwriting excisting data.

Do you agree with me on that and would be willing to add this behavior to chatscript?

**Comment by bwilcox-1234** [comment]
The problem is, if it cannot open the file eventually, then it cannot respond to the user. A single retry is perfectly reasonable. In non-server mode it could afford to stall for a 1 second and then retry, in server mode it would be immediate

**Comment by Tobi-La** [comment]
Jeah, maybe a single retry is all I need. That would definitely be an improvement.
If the retry doensn't help, I would prefer an empty response over overwriting, but one retry is better then none.

So this will be updated in future versions? Then I would not implement a workaround yet and wait for the change first.

Thank you very much for your help Bruce :)

**Comment by bwilcox-1234** [comment]
I will do single retry in future version. I can also test the return code and if it is appropriate, return a fixed message of "Huh?"

**Comment by Tobi-La** [comment]
Jeah, a fixed message would be fine for me :)
